### PR TITLE
DEV: add a conda environment.yml with all development dependencies

### PR DESCRIPTION
**Pull request by rgommers**
There are enough dependencies now that it's too hard to remember. This set is comprehensive, so a single
`conda env create -f environment.yml` gives you everything in a fresh environment named `numpy-dev`.

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# To use:
+#   $ conda env create -f environment.yml  # `mamba` works too for this command
+#   $ conda activate numpy-dev
+name: numpy-dev
+channels:
+  - conda-forge
+dependencies:
+  - python
+  - cython
```
Comment (mattip) [inline environment.yml:9]:
```suggestion
  - cython=0.29.22
```

File: environment.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  - compilers
+  - openblas
```
Comment (mattip) [inline environment.yml:11]: hmm. Shouldn't this be using our version of OpenBLAS?

File: environment.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  - mypy=0.812
+  - typing_extensions
+  # For building docs
+  - sphinx=3.5.2
+  - numpydoc=1.1.0
+  - ipython
+  - scipy
+  - pandas
+  - matplotlib
+  - pydata-sphinx-theme
```
Comment (mattip) [inline environment.yml:27]:
```suggestion
  - pydata-sphinx-theme-0.5.0
```

File: environment.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  # For linting
+  - pycodestyle=2.7.0
+  - gitpython
+  # Used in some tests
+  - cffi
+  - pytz
+  - pickle5
```
Comment (mattip) [inline environment.yml:34]: pickle5 for python<=3.7 and implementation != PyPy

**Review by mattip — COMMENTED** [review]
Using conda could be mentioned in the documentation:
- for [building docs](https://numpy.org/devdocs/docs/howto_build_docs.html#instructions)
- for [building for tests and docs](https://numpy.org/devdocs/dev/index.html#test-coverage)

and maybe even describe conda in [describing virtualenvs](https://numpy.org/devdocs/dev/development_environment.html#using-virtualenvs)

How do we keep the versions synced with the requirements.txt files? Does dependabot see this file?

File: environment.yml (hunk 1 above)
Comment (rgommers) [reply environment.yml:9]: No, I only want to pin things that are very likely to cause problems. This is unnecessary, and also wrong in the corresponding requirements.txt file.

File: environment.yml (hunk 2 above)
Comment (rgommers) [reply environment.yml:11]: No, it should use the conda-forge provided one.

File: environment.yml (hunk 3 above)
Comment (rgommers) [reply environment.yml:27]: Also no, don't want to pin this.

File: environment.yml (hunk 4 above)
Comment (rgommers) [reply environment.yml:34]: Maybe better to leave it out then, it's a dont-care anyway.

**Comment by rgommers** [comment]
> and maybe even describe conda in [describing virtualenvs](https://numpy.org/devdocs/dev/development_environment.html#using-virtualenvs)

I think that's the one place I'd like to update.

> How do we keep the versions synced with the requirements.txt files? Does dependabot see this file?

Dependabot doesn't see it AFAIK. There's only a few pins, which are easy enough to update by hand if any test fails. I don't expect that to happen often, probably only once a year or less.

**Comment by bashtage** [comment]
Should this be included in CI to make sure it doesn't get out of date?

**Comment by rgommers** [comment]
> Should this be included in CI to make sure it doesn't get out of date?

I think that's a good idea, however I'd like to file a follow-up issue for that. Adding an environment file is a useful enhancement in itself, and I don't have the bandwidth to switch some CI jobs to conda right now.

File: environment.yml (hunk 1 above)
Comment (charris) [reply environment.yml:9]: A more explicit name would help, maybe `conda-environment.yml`?

Comment (rgommers) [reply environment.yml:9]: `environment.yml` is the standard name for this file in other projects and in educational material (example: [SW Carpentry](https://carpentries-incubator.github.io/introduction-to-conda-for-data-scientists/04-sharing-environments/index.html)). It's also what IDEs and tools like GitHub will look for if they gain conda support. So renaming is not a good idea.

Analogy: you wouldn't call something `pypi_requirements.txt` either.

File: environment.yml (hunk 4 above)
Comment (rgommers) [reply environment.yml]: I removed this, resolving.

File: environment.yml (hunk 1 above)
Comment (rgommers) [reply environment.yml:9]: Pandas equivalent: https://github.com/pandas-dev/pandas/blob/master/environment.yml

**Comment by charris** [comment]
This could use a release note, probably under `new_feature`.  I wonder if this and the various `*_requirements.txt` files should be included in the source distribution. We include `runtests.py` and `tox.ini`.

**Comment by rgommers** [comment]
> This could use a release note, probably under `new_feature`. I wonder if this and the various `*_requirements.txt` files should be included in the source distribution? We include `runtests.py` and `tox.ini`.

I think yes, they should. Not that they'll be used much that way (the only way to do so is to manually download the tarball and go from there), but an sdist is also a snapshot of a release with archiving/historical value. I'll update this.

**Comment by rgommers** [comment]
I fixed the `MANIFEST.in` includes. `setup.cfg` was missing too. Changed to `include *.*` followed by excluding things, that way if we make a mistake we include a little too much which is better than including too little.

Decided against a release note. This is not a user-accessible feature from a release artifact, so it only belongs in the developer docs.

**Comment by charris** [comment]
Thanks Ralf.

**Pull request merged**